lib/api/statistics.py

```diff
@@ -0,0 +1,55 @@
+from datetime import datetime, timedelta
+
+from rest_framework import status
+from rest_framework import generics
+from django.db.models import Count, QuerySet
+from django.utils import timezone
+
+from exercise.submission_models import Submission
+from .serializers import StatisticsSerializer
+
+
+class BaseStatisticsView(generics.RetrieveAPIView):
+    """
+    Returns submission statistics for the entire system, over a given time window.
+
+    Returns the following attributes:
+
+    - `submission_count`: total number of submissions.
+    - `submitters`: number of users submitting.
+
+    Operations
+    ----------
+
+    `GET /statistics/`:
+        returns the statistics for the system.
+
+    - URL parameters:
+        - `endtime`: date and time in ISO 8601 format indicating the end point
+          of time window we are interested in. Default: now.
+        - `starttime`: date and time in ISO 8601 format indicating the start point
+          of time window we are interested in. Default: one day before endtime
+    """
+    serializer_class = StatisticsSerializer
+
+    def get_queryset(self) -> QuerySet:
+        queryset = Submission.objects.all()
+
+        endtime = self.request.query_params.get('endtime')
+        starttime = self.request.query_params.get('starttime')
+        serializer = StatisticsSerializer(data={'starttime': starttime, 'endtime': endtime})
```

Comment:
Since you have hardcoded `StatisticsSerializer` here instead of calling `self.get_serializer(data={'starttime': starttime, 'endtime': endtime})`, then the child view classes actually also use `StatisticsSerializer` even though those classes declare another serializer class in the class variables. This will cause confusion since this does not behave in the usual Django REST framework way.

When you change this, you will probably need to add `read_only=True` to the child serializers (`ExerciseStatisticsSerializer` and `CourseStatisticsSerializer`).

Reply:
done